central/processindicator/datastore/datastore_sac_test.go

```diff
@@ -0,0 +1,571 @@
+		s.indexer = index.New(s.index)
+	}
+
+	s.boltDb = testutils.DBForSuite(s)
+	s.commentsStorage = commentsstore.New(s.boltDb)
+	s.search = search.New(s.storage, s.indexer)
+	s.datastore, err = New(s.storage, s.commentsStorage, s.indexer, s.search, nil)
+	s.Require().NoError(err)
+
+	s.testContexts = sacTestUtils.GetNamespaceScopedTestContexts(context.Background(), s.T(),
+		resources.Indicator.GetResource())
+}
+
+func (s *processIndicatorDatastoreSACSuite) TearDownSuite() {
+	if features.PostgresDatastore.Enabled() {
+		s.pool.Close()
+	} else {
+		err := rocksdb.CloseAndRemove(s.engine)
+		s.Require().NoError(err)
+	}
+
+	testutils.TearDownDB(s.boltDb)
+
+	s.Require().NoError(s.index.Close())
+}
+
+func (s *processIndicatorDatastoreSACSuite) SetupTest() {
+	s.testProcessIndicatorIDs = make([]string, 0)
+
+	processIndicators := fixtures.GetSACTestProcessIndicatorSet()
+	err := s.datastore.AddProcessIndicators(s.testContexts[sacTestUtils.UnrestrictedReadWriteCtx], processIndicators...)
+	s.Require().NoError(err)
+
+	for _, pi := range processIndicators {
+		s.testProcessIndicatorIDs = append(s.testProcessIndicatorIDs, pi.GetId())
+	}
+}
+
+func (s *processIndicatorDatastoreSACSuite) TearDownTest() {
+	err := s.datastore.RemoveProcessIndicators(s.testContexts[sacTestUtils.UnrestrictedReadWriteCtx],
+		s.testProcessIndicatorIDs)
+	s.Require().NoError(err)
+}
+
+func (s *processIndicatorDatastoreSACSuite) deleteProcessIndicator(id string) {
+	s.Require().NoError(s.datastore.RemoveProcessIndicators(s.testContexts[sacTestUtils.UnrestrictedReadWriteCtx],
+		[]string{id}))
+}
+
+func (s *processIndicatorDatastoreSACSuite) TestAddProcessIndicators() {
+	cases := map[string]struct {
+		scopeKey    string
+		expectFail  bool
+		expectedErr error
+	}{
+		"global read-only should not be able to add": {
+			scopeKey:    sacTestUtils.UnrestrictedReadCtx,
+			expectFail:  true,
+			expectedErr: sac.ErrResourceAccessDenied,
+		},
+		"global read-write should be able to add": {
+			scopeKey: sacTestUtils.UnrestrictedReadWriteCtx,
+		},
+		"read-write on wrong cluster should not be able to add": {
+			scopeKey:    sacTestUtils.Cluster1ReadWriteCtx,
+			expectFail:  true,
+			expectedErr: sac.ErrResourceAccessDenied,
+		},
+		"read-write on wrong cluster and namespace should not be able to add": {
+			scopeKey:    sacTestUtils.Cluster1NamespaceAReadWriteCtx,
+			expectFail:  true,
+			expectedErr: sac.ErrResourceAccessDenied,
+		},
+		"read-write on wrong cluster and matching namespace should not be able to add": {
+			scopeKey:    sacTestUtils.Cluster1NamespaceBReadWriteCtx,
+			expectFail:  true,
+			expectedErr: sac.ErrResourceAccessDenied,
+		},
+		"read-write on matching cluster and wrong namespace should not be able to add": {
+			scopeKey:    sacTestUtils.Cluster2NamespaceAReadWriteCtx,
+			expectFail:  true,
+			expectedErr: sac.ErrResourceAccessDenied,
+		},
+		"read-write on matching cluster and matching namespace should be able to add": {
```

Comment:
Can you add a test for write permission for whole matching cluster ?

Reply:
By `whole cluster` do you mean something similar to the existing `read-write on matching cluster can read` ? Otherwise, can't really follow what you mean 😞

Reply:
yes, basically a test with `Cluster2ReadWriteCtx` as scopeKey